File: Problems/No.74_Search_a_2D_Matrix/search_a_2D_matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """

        # parameters
        rowS, colS = len(matrix), len(matrix[0])
        left, right, mid = 0, rowS*colS-1, None
        r, c = None, None

        # start computing
        while (left <= right):
            
            mid = left + (right-left)//2
            r = mid // colS
            c = mid % colS
                 
            if (matrix[r][c] == target):
                return True
            elif (matrix[r][c] < target):
                left = mid + 1
            else:
                right = mid - 1
        
        return False

File: Problems/No.74_Search_a_2D_Matrix/search_a_2D_matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        // parameters
        int rowS = matrix.size(), colS = matrix[0].size();
        int left = 0, right = rowS*colS-1, mid;
        int r, c;

        // start computing (use Binary Search)
        while (left <= right) {

            mid = left + (right - left)/2;
            r = mid / colS;
            c = mid % colS;
            
            if (matrix[r][c] == target) return true;
            else if (matrix[r][c] < target) left = mid + 1;
            else right = mid - 1;

        }

        return false;

    }
};
